### ISSUE TITLE
Surface Go 3 Typecover can't be configured under libinput without USB_QUIRK_DELAY_INIT or rebinding

### ISSUE THREAD

**Issue by WolfWings** [linux-surface/linux-surface#1059]
The short version: LibInput doesn't initialize the touchpad properly without this quirk, preventing sway's input configuration to change various settings (dwt, clickfingers, etc) from being applied unless I unbind/bind the typecover (or any equivalent of that like physically detaching and reattaching it) at which point libinput configuration is properly applied.

Applying the quirk via `usbcore.quirks=045e:09b5:g` corrects this issue without requiring the rebinding.

This may be a LibInput race condition? I don't know how to rule that out and this quirk resolves things so I thought best to post this issue as it may address various other touchpad/typecover-related issues I'm seeing.

### Environment

System is configured to auto-login on tty1 as the `tablet` user, which has their .bashrc configured to start `sway` when logged in on tty1, which in turn is configured to lock the session on initial startup. It's a tablet only intended to be used by a single user, and mostly as a Krita tool with a browser. This setup allows for using other authentication methods such as pin/pattern-based which work better on the touchscreen when carrying the tablet one-handed, etc.

I tried all of `*`, `type:touchpad`, `1118:2485:Microsoft_Surface_Keyboard_Touchpad` for configuring the touchpad using the standard `sway-input` options for libinput, it made no difference. Rebinding or the quirk suddenly the configuration is applied properly. Using `swaymsg` to manually send configurations report they apply in the debug log but don't take effect.

Some limited testing under x.org w/ the `synaptics` driver instead appears to configure correctly, but that takes a second or two longer to initialize when I looked at the logs so it might merely be mimicking this quirk by accident. Under x.org libinput behaved identically with the touchpad ignoring all settings changes without rebinding somehow.

- Hardware model: Surface Go 3
- Kernel version: Linux ~~private~~ 6.1.11-arch1-1-surface #<span></span>1 SMP PREEMPT_DYNAMIC Sat, 11 Feb 2023 22:53:48 +0000 x86_64 GNU/Linux
- Distribution: Arch Linux

As attaching the dual dmesg outputs resulted in a 'comment too long' error, I'm attaching them instead.
[dmesg_WITH_quirk.txt](https://github.com/linux-surface/linux-surface/files/10764587/dmesg_WITH_quirk.txt)
[dmesg_WITHOUT_quirk.txt](https://github.com/linux-surface/linux-surface/files/10764588/dmesg_WITHOUT_quirk.txt)

**Comment by WolfWings** [comment]
Also as I have several items on my kernel command line:
- `acpi_os_name="Windows 2022"` is to match the original hardware coming with Windows 11 instead of reporting Windows 8 as I was testing if it changed some ACPI results. It did not, and removing it doesn't change the touchpad behavior I'm reporting. I can re-upload fresh dmesg outputs if needed.
- `nmi_watchdog=0` to shut up the powertop warning
- `8250.nr_uarts=0` because for some reason the linux kernel generates 32 serial port entries by default on the Go 3 and SystemD takes a few seconds attempting to enumerate all of them and failing before boot continues
- `clocksource=tsc` also to speed up the boot because the kernel bounced through multiple clock sources by default due to its default probing order and waits for each to stabilize otherwise
- `fbcon=font:TER16x32` to use the built-in 16x32 font instead of installing console fonts from userspace

**Comment by qzed** [comment]
> This may be a LibInput race condition? I don't know how to rule that out and this quirk resolves things so I thought best to post this issue as it may address various other touchpad/typecover-related issues I'm seeing in other issues.

Unfortunately I don't know whether that's a possibility. It's a bit odd that the typecover works without that quirk and it's just the configuration part that's the issue. So I guess it does somewhat point to an issue in user-space. I also don't know whether there's some communication going back to the touchpad device as part of that configuration step that could be ignored by the device if it was sent too early. On the other hand: If the quirk reliably fixes this, it may very well be an underlying issue with the kernel/device, and that should be addressed in the kernel.

The libinput developers might know more about that, so you might want to talk to those.

Anyways, below is a diff applying that quirk in case we decide that it is a kernel issue. I guess you could also send this as a patch to the linux-usb mailing list (and respective maintainers, use e.g. `scripts/get_maintainer.sh` in the kernel source) and bring up the issue there. Ultimately it's up to them to decide if this should be handled in the kernel.

```diff
diff --git a/drivers/usb/core/quirks.c b/drivers/usb/core/quirks.c
index 079e183cf3bf..eebf32a93786 100644
--- a/drivers/usb/core/quirks.c
+++ b/drivers/usb/core/quirks.c
@@ -220,6 +220,9 @@ static const struct usb_device_id usb_quirk_list[] = {
        /* Microsoft Surface Dock Ethernet (RTL8153 GigE) */
        { USB_DEVICE(0x045e, 0x07c6), .driver_info = USB_QUIRK_NO_LPM },
 
+       /* Microsoft Surface Go 3 Type-Cover */
+       { USB_DEVICE(0x045e, 0x09b5), .driver_info = USB_QUIRK_DELAY_INIT },
+
        /* Cherry Stream G230 2.0 (G85-231) and 3.0 (G85-232) */
        { USB_DEVICE(0x046a, 0x0023), .driver_info = USB_QUIRK_RESET_RESUME },
 
```

**Comment by qzed** [comment]
Any chance you're experiencing a similar issue as in #305 without the quirk (i.e. multi-touch stops working)? In that case the quirk should definitely be applied in the kernel.

**Comment by WolfWings** [comment]
@qzed If you mean multi-touch on the touchpad? Yes.

I can't use any multi-finger anything, no clicks, no scrolling, only the 'defined zones' work (edge scroll, fixed 'button' zones for middle/right) without the rebind.

I'd thought it was a libinput related issue and approached it as such until I found the quirk resolved it, but appears to be #305 now that I see that bug. The evtest confirms, nothing works for multi-finger on the touchpad at all without the quirk/rebind.

**Comment by qzed** [comment]
> Also just saw the comment in the other issue regarding the pointer speed being much higher, that matches what I'm seeing as well.

Yeah, makes sense that when it's falling back to the "mouse emulation" mode the pointer settings like speed and everything will be different. So I we probably want to apply that quirk to the type-covers of both the Go 2 and 3. Maybe even the one of the Go 1 and other USB-based type-covers.

**Comment by WolfWings** [comment]
Do you need me to puzzle out how to poke the linux-usb ML? Or are you folks in contact when them at some points in general and can drive that part?

**Comment by qzed** [comment]
I think we've confirmed that it should go into the kernel since the kernel switches between the modes, so I don't think there's much need to bug them about that. That means we have two ways forward:
- You write a patch, submit it to the ML and send me a link or PR so that I can pull it into our kernel. Pros: You get all the credit and can brag about it :smiley:. Cons: You need to do a bit of work and might need to set up a mail for submitting patches via `git send-email`.
- I do the same thing (although I'd probably wait for some more feedback before submitting it upstream, especially whether we can include the Go 2 or more devices on that). Pros: You don't need to do anything. Cons: You get slightly less credit (I'll still link this thread here and can credit you with `Reported-by: Your Name <your email>`, if you want that).

**Comment by WolfWings** [comment]
I'm fine with the latter, that seems a better fit for the existing linux-surface workflow I see here to me of getting things sorted out here then pushing them upstream once details have been puzzled out like extra devices to (not) include in the quirk, etc.

Especially since you already wrote the patch and it's more patch format boilerplate than actual code change. :)

**Comment by qzed** [comment]
The latest kernels should now contain this quirk for the Go 3. I still want to wait a bit and see if the Go 2 or other devices might need it as well, and also collect some feedback on the other type-cover issues.

**Comment by WolfWings** [comment]
Awesome to hear, thank you! And hopefully it turns out other USB-based typecover issues might end up resolved with this.
